File: c++/cote/GJM/삼각형_최대경로.cpp
#include <iostream>
#include <random>
#define N 10
#define MAX_VALUE 15
#define TEST_CASES 10

using namespace std;

int** initialize(default_random_engine& gen){
	uniform_int_distribution<int> dis(1, MAX_VALUE);

	int** map = new int*[N];

	for(int i =0; i<N; i++){
		map[i] = new int[N];
	}


	for(int i =0; i< N; i++){
		for(int j=0; j<= N; j++){
			map[i][j] = dis(gen);
		}
	}

	return map;
}


int path(int x, int y, int sum, int** map){
	if(y == N-1) return sum + map[y][x];
	sum += map[y][x];
	return max(path(x,y+1,sum,map), path(x+1,y+1,sum, map));
}

int solve(int** map){

	return path(0,0,0,map);
}


int path2(int x, int y, int sum, int** map, int cache[N][N][MAX_VALUE*100+1]){
	
	if(y == N-1) return sum + map[y][x];
	int& ret = cache[y][x][sum];
	if( ret != -1) return ret;
	sum += map[y][x];
	return (ret = max(path2(x,y+1,sum,map,cache), path2(x+1,y+1,sum, map,cache)));
}

int solve2(int** map){
	int cache[N][N][MAX_VALUE*100+1];
	memset(cache, -1, sizeof(cache));

	return path2(0,0,0,map,cache);
}

int path3(int x, int y, int** map, int cache[N][N]){
	
	if(y == N-1) return map[y][x];
	int& ret = cache[y][x];
	if( ret != -1) return ret;
	return (ret = max(path3(x,y+1,map,cache), path3(x+1,y+1, map,cache)) + map[y][x]);
}

int solve3(int** map){
	int cache[N][N];
	memset(cache, -1, sizeof(cache));

	return path3(0,0,map,cache);
}

int main(void){
	default_random_engine gen(43);

	for(int i =0; i<TEST_CASES; i++){

		int** map = initialize(gen);


		cout << solve3(map) << endl;

		for(int i=0; i<N; i++){
			delete[] map[i];
		}
		delete[] map;
	}


	return 0;
}

File: javascript/cote/GJM/삼각형_최대경로.mjs
const N = 10
const MAX_VALUE = 15
const TEST_CASE = 10

let cache;

function initialize() {
	const ret = []

	for (let i = 0; i < N; i++) {
		ret.push([])
		for (let j = 0; j<= i; j++) {
			ret[i].push(Math.floor(Math.random() * MAX_VALUE))
		}
	}

	return ret
}

function search(y, x, map, cache) {
	if (y == N - 1){
		return map[y][x]
	}

	let ret = cache[y][x]

	if (ret != -1) {
		return ret
	}

	ret = 0

	ret = Math.max(ret, search(y + 1, x, map, cache) + map[y][x])
	ret = Math.max(ret, search(y + 1, x + 1, map, cache) + map[y][x])

	return ret

}

function solve(map) {
	cache = new Array(N + 1).fill(0).map(_ => new Array(N + 1).fill(-1))

	return search(0, 0, map, cache)
}

for (let it = 0; it < TEST_CASE; it++) {
	const map = initialize()
	console.log(solve(map))
}